feat(ui/logs)_#47619: add toggle for log grouping

davidfgcorreia opened the pull request:
- Add button in header to enable or disable log grouping
- Allow logs to be shown linearly or grouped
- Adjust group offset and display
https://www.youtube.com/watch?v=doCR4YM1Rek

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #47619 
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

davidfgcorreia committed:
feat(ui/logs)_#47619: add toggle for log grouping

- Add button in header to enable or disable log grouping
- Allow logs to be shown linearly or grouped
- Adjust group offset and display

Co-authored-by: Francisco Núncio <francisco.nuncio@tecnico.ulisboa.pt>

jscheffl commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -189,6 +193,18 @@ export const TaskLogHeader = ({
             </Button>
           </Tooltip>
+          {toggleNested ? (
+            <Button
```
Nit: Button distance is un-aligned right and left from space. Can you please make this the same?
![image](https://github.com/user-attachments/assets/698b48b0-be06-4851-8fcd-ae531f3bd075)

davidfgcorreia replied:
![imagem_2025-05-29_132549424](https://github.com/user-attachments/assets/cf91e03d-cd72-4a06-8f31-30eb144640be)

Distances ajusted.

jscheffl replied:
Thanks for fixing the alignment. Unfortuantely I am a bit picky about consistency and pixels... now the button distance is different to the buttons on top. Can you adjust for consistency?
![image](https://github.com/user-attachments/assets/3dbf4c16-36c8-4b12-aba8-0b6737ed1992)

davidfgcorreia replied:
![imagem_2025-05-29_171812084](https://github.com/user-attachments/assets/cd54a45a-4001-4cb5-b853-9869c3cd35d6)
 I think i got it rigth this time.

jscheffl replied:
Thanks for the rework! Ähm... oh. I am really sorry but now (I hope the last) nitpick is the spacing inside the buttons (text distance left/right to the button border) is different if you compare the buttons above and the lower ones.

![image](https://github.com/user-attachments/assets/dc3fbd3d-e10e-4a6c-877f-cbc7b96ec447)

Can I kindly ask to normalize this as well?

davidfgcorreia replied:
![imagem_2025-05-31_180904536](https://github.com/user-attachments/assets/5aad151e-3404-4717-8150-deaea9c71259)
Additionally, the button's background color has been updated.

jscheffl reviewed (commented):
Looks good for me. Would propose to merge once the small nit is resolved.

davidfgcorreia committed:
Fix Button distance

davidfgcorreia committed:
ajust gap betewn butons

davidfgcorreia committed:
feat(ui/logs): logs UX, keyboard accessibility, and code quality

- Add expand/collapse all functionality for log groups using the "expanded" prop
- Ensure initial logs state is collapsed and button shows "Expand All"
- Add keyboard shortcut (E) to toggle expand/collapse all log groups
- Unify button styles with black background and consistent padding/hover

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -111,6 +115,24 @@ export const TaskLogHeader = ({
+      globalThis.removeEventListener("keydown", handler);
+    };
+  }, [toggleExpanded]);
+
```
You can just check how we call `useHotkeys()` in other places

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -178,20 +200,57 @@ export const TaskLogHeader = ({
+                size="md"
+                variant="outline"
+              >
+                {expanded ? "Collapse All" : "Expand "}
```
I think its also ok to reuse the expand all / collapse all icons we use for the grid view in `ToggleGroups.tsx`

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -178,20 +200,57 @@ export const TaskLogHeader = ({
             <Tooltip closeDelay={100} content="Press f for fullscreen" openDelay={100}>
-              <IconButton aria-label="Full screen" bg="bg.panel" onClick={toggleFullscreen} variant="outline">
+              <IconButton
+                _hover={{ bg: "gray.800" }}
```
Let's use the semantic colors here. 
```suggestion
                _hover={{ bg: "bg.emphasized" }}
```

davidfgcorreia replied:
Also changed the warp buton to this houver style

bbovenzi reviewed (commented):
A few nits and then we should update the existing tests and add a new one for a nested log group

davidfgcorreia committed:
feat(ui/logs): unify expand/collapse log controls with grid view style

-Rebase
- Refactor TaskLogHeader to use a ButtonGroup with IconButton for expand/collapse,
  matching the style and behavior of ToggleGroups in the grid view
- Warnings on precomit but not on the file i changed

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/Logs.tsx:
```diff
@@ -94,17 +100,19 @@ export const Logs = () => {
       {showExternalLogRedirect && externalLogName && taskInstance ? (
         tryNumber === undefined ? (
-          <p>No try number</p>
+          <p>{translate("No try number")}</p>
```
Let's rebase and use the correct translate here.

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -178,20 +183,57 @@ export const TaskLogHeader = ({
+          <Tooltip closeDelay={100} content="Press e to expand/collapse" openDelay={100}>
+            <ButtonGroup attached size="md" variant="outline">
+              <IconButton
+                aria-label="Expand"
```
The aria-labels can also call `translate()`

pierrejeambrun replied:
has been done

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -178,20 +183,57 @@ export const TaskLogHeader = ({
               {wrap ? "Unwrap" : "Wrap"}
             </Button>
           </Tooltip>
+          <Tooltip closeDelay={100} content="Press e to expand/collapse" openDelay={100}>
```
Let's translate the tooltip content

pierrejeambrun replied:
has been done

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Logs/TaskLogHeader.tsx:
```diff
@@ -178,20 +183,57 @@ export const TaskLogHeader = ({
               aria-label={wrap ? "Unwrap" : "Wrap"}
-              bg="bg.panel"
+              bg="black"
+              color="white"
```
Let's remove these or pass `"bg"` and `"fg"` so that it isnt black in light mode too

pierrejeambrun replied:
done

bbovenzi reviewed (commented):
Let's rebase and manage some translations. Otherwise looking good

davidfgcorreia committed:
feat(i18n/ui): add expand/collapse translations, fix requested issues, and rebase

- Added "expand" section with tooltip, expand, and collapse strings to all locale JSON files
- Fixed requested issues in the UI related to expand/collapse controls
- Performed rebase to update branch with latest changes

pierrejeambrun commented on airflow-core/src/airflow/ui/src/i18n/locales/de/common.json:
```diff
@@ -63,6 +63,11 @@
+    "collapse": "Einklappen",
+    "expand": "Erweitern",
+    "tooltip": "Drücken Sie {{hotkey}}, um zu erweitern"
+  },
```
Just fill the english translations (because those are the one we can approve) and let the other translation owners update their respective files in follow up PRs.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -95,40 +94,72 @@ const parseLogs = ({
+
+        if (finishedGroup) {
+          const groupElement = (
+            <Box key={finishedGroup.name + Math.random()} mb={2} pl={finishedGroup.level * 2}>
```
Don't push a random number on the key, that defeats the purpose keys and react diffing between renders.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -95,40 +94,72 @@ const parseLogs = ({
+
+      if (unfinished) {
+        result.push(
+          <Box key={unfinished.name + Math.random()} mb={2} pl={unfinished.level * 2}>
```
Same

pierrejeambrun reviewed (commented):
Nice PR thanks, just a few comments in addition to other things mentioned by Brent and Jens.

Looking good and working as expected.

pierrejeambrun reviewed (approved):
Looking good, CI need fixing, one test needs to be adjusted because by default groups are collapsed.

guan404ming commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -95,40 +94,72 @@ const parseLogs = ({
+        if (finishedGroup) {
+          const groupElement = (
+            <Box key={finishedGroup.name} mb={2} pl={finishedGroup.level * 2}>
+              <details open={open} style={{ width: "100%" }}>
```
```suggestion
              <chakra.details open={open} w="100%">
```

to use [chakra factory](https://chakra-ui.com/docs/styling/chakra-factory) could preserve consistency and get better support for RTL lang

guan404ming commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -95,40 +94,72 @@ const parseLogs = ({
+          const groupElement = (
+            <Box key={finishedGroup.name} mb={2} pl={finishedGroup.level * 2}>
+              <details open={open} style={{ width: "100%" }}>
+                <summary data-testid={`summary-${finishedGroup.name}`}>
```
```suggestion
                <chakra.summary data-testid={`summary-${finishedGroup.name}`}>
```
same as above

guan404ming reviewed (approved):
LGTM, 2 nits.

pierrejeambrun reviewed (commented):
beside @guan404ming small suggestions, It's almost ready, it would be great if you could also fix the last remaining test @davidfgcorreia so we can merge this :), thanks for your work.

davidfgcorreia committed:
feat(i18n/ui): add expand/collapse translations, improve log header i18n, and update environment typings

Updated log header UI (TaskLogHeader.tsx) to use i18n for expand/collapse/fullscreen tooltips and aria-labels.
Fixed and improved expand/collapse controls and tooltips in the UI for accessibility and localization.
Updated vite-env.d.ts to ensure correct Vite and ImportMeta type references for the Airflow UI.
Minor refactors and cleanup in log and query files to support new i18n keys and UI consistency.

davidfgcorreia committed:
fix(ui/logs): restore log grouping UI and update related test

Changes to the log grouping UI in useLogs.tsx, using <chakra.details> and <chakra.summary> for grouped log sections.
Ensures log groups are rendered with correct expand/collapse behavior and accessibility.
Updates the related test (Logs.test.tsx) to match the new log grouping and expand/collapse structure.

davidfgcorreia committed:
test(logs): adjust test selectors to handle multiple collapse/expand buttons

pierrejeambrun reviewed (approved):
LGTM
